lib/SILOptimizer/Transforms/Outliner.cpp

```diff
@@ -0,0 +1,1306 @@
+    // Move the release after the call.
+    Release->moveBefore(StartBB->getTerminator());
+  }
+
+  //     [StartBB]
+  //    /        \
+  // [NoneBB]  [SomeBB]
+  //   \          /
+  //   [OldMergeBB]
+  //
+  //   Split to:
+  //
+  //      [StartBB]
+  //          |
+  //   [OutlinedEntryBB]   }
+  //    /        \         }
+  // [NoneBB]  [SomeBB]    } outlined
+  //   \          /        }
+  //   [OldMergeBB]        }
+  //       |
+  //   [NewTailBB]
+  //
+  auto *OutlinedEntryBB = StartBB->split(SILBasicBlock::iterator(FirstInst));
+  auto *OldMergeBB = Br->getDestBB();
+  auto *NewTailBB = OldMergeBB->split(OldMergeBB->begin());
+
+  // Call the outlined function.
+  {
+    SILBuilder Builder(StartBB);
+    auto Loc = FirstInst->getLoc();
+    SILValue FunRef(Builder.createFunctionRef(Loc, Fun));
+    SILValue Apply(
+        Builder.createApply(Loc, FunRef, {FirstInst->getOperand(0)}, false));
+    Builder.createBranch(Loc, NewTailBB);
+    OldMergeBB->getArgument(0)->replaceAllUsesWith(Apply);
+  }
+
+  if (!NeedsDefinition) {
+    // Delete the outlined instructions/blocks.
+    if (Release)
+      Release->eraseFromParent();
+    OutlinedEntryBB->eraseInstructions();
+    OutlinedEntryBB->eraseFromParent();
+    NoneBB->eraseInstructions();
+    NoneBB->eraseFromParent();
+    SomeBB->eraseInstructions();
+    SomeBB->eraseFromParent();
+    OldMergeBB->eraseInstructions();
+    OldMergeBB->eraseFromParent();
+    return std::make_pair(nullptr, std::prev(StartBB->end()));
+  }
+
+  if (OutlinedEntryBB->getParent()->hasUnqualifiedOwnership())
+    Fun->setUnqualifiedOwnership();
+
+  Fun->setInlineStrategy(NoInline);
+
+  // Move the blocks into the new function.
+  auto &FromBlockList = OutlinedEntryBB->getParent()->getBlocks();
+  Fun->getBlocks().splice(Fun->begin(), FromBlockList, OldMergeBB);
+  Fun->getBlocks().splice(Fun->begin(), FromBlockList, NoneBB);
+  Fun->getBlocks().splice(Fun->begin(), FromBlockList, SomeBB);
+  Fun->getBlocks().splice(Fun->begin(), FromBlockList, OutlinedEntryBB);
+
+  // Create the function argument and return.
+  auto *Load = dyn_cast<LoadInst>(FirstInst);
+  SILBuilder Builder(FirstInst);
+  if (Load) {
+    OutlinedEntryBB->createFunctionArgument(Load->getOperand()->getType());
+    auto *NewLoad =
+        Builder.createLoad(Load->getLoc(), OutlinedEntryBB->getArgument(0),
+                           Load->getOwnershipQualifier());
+    Load->replaceAllUsesWith(NewLoad);
+    Load->eraseFromParent();
+  } else {
+    OutlinedEntryBB->createFunctionArgument(
+        FirstInst->getOperand(0)->getType());
+    auto *Arg = OutlinedEntryBB->getArgument(0);
+    FirstInst->setOperand(0, Arg);
+    PropApply->setArgument(0, Arg);
+  }
+  Builder.setInsertionPoint(OldMergeBB);
+  Builder.createReturn(ClassMethod->getLoc(), OldMergeBB->getArgument(0));
+  return std::make_pair(Fun, std::prev(StartBB->end()));
+}
+
+#define ADVANCE_ITERATOR_OR_RETURN_FALSE(It)                                   \
```

Comment:
There is something in STLExtras similar to this called next_or_end. Just an FYI. But I guess you are using the macro so you can return easier?

Regardless, shouldn't this be wrapped in a:

```
do { ... } while (0);
```

Reply:
I use this macro to make the return on end() easier.